pkg/operators/v1alpha1/catalogsource_types.go

```diff
@@ -242,6 +243,28 @@ func (c *CatalogSource) Poll() bool {
 	return true
 }
 
+// UnmarshalJSON implements the encoding/json.Unmarshaler interface.
+func (u *UpdateStrategy) UnmarshalJSON(data []byte) (err error) {
```

Comment:
It's not immediately clear to me why we'd need to build up this kind of validation check - it's possible I'm missing how we'd use this in OLM. Would it be possible to take a CatalogSource as input, and validate that the `.spec.RegistryPoll.Interval` matches the type signature and format we're expecting? Is this a candidate to enforcing this fields' schema validation such that only time values are acceptable? I'm not sure whether the latter would be a breaking change.

Reply:
Hi Tim,

This PR is meant to solve this issue -> https://bugzilla.redhat.com/show_bug.cgi?id=1943937

> It's not immediately clear to me why we'd need to build up this kind of validation check

We need to build up this validation check because the operator cannot handle invalid interval values.

We see error in the operator logs:
```
reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.CatalogSource: v1alpha1.CatalogSourceList.Items: []v1alpha1.CatalogSource: v1alpha1.CatalogSource.v1alpha1.CatalogSource.Spec: v1alpha1.CatalogSourceSpec.UpdateStrategy: v1alpha1.UpdateStrategy.RegistryPoll: v1alpha1.RegistryPoll.Interval: unmarshalerDecoder: time: unknown unit "mCopy code" in duration "45mError code", error found in #10 byte of ...|Copy code"}}}},{"api|..., bigger context ...|rategy":{"registryPoll":{"interval":"45mCopy code"}}}},{"apiVersion":"operators.coreos.com/v1alpha1"|...
```

> Would it be possible to take a CatalogSource as input, and validate that the .spec.RegistryPoll.Interval matches the type signature and format we're expecting

Is there an advantage to this approach?

> Is this a candidate to enforcing this fields' schema validation such that only time values are acceptable?

The user can still put any value such as:

```
  updateStrategy:
    registryPoll:
      interval: 45mError code    ################### incorrect value
```

But since it is invalid, it would get converted to default interval value.

> I'm not sure whether the latter would be a breaking change.

Adding schema validation would be a breaking change.

Conversation related to this could be found here -> https://coreos.slack.com/archives/G3T7N42NP/p1625153985037600?thread_ts=1625153575.035700&cid=G3T7N42NP

Reply:
Yeah, that's true, adding extra schema validation around this field could be viewed as a breaking change. I think I still need to wrap my head around this usage in OLM, and would prefer an approach where we validate a CatalogSource as an input vs. serializing and parsing nested fields: https://github.com/operator-framework/api/blob/master/pkg/validation/internal/csv.go#L32. I'll poke around that slack conversation later :+1:.